dm/dm/worker/source_worker.go

```diff
@@ -1097,3 +1121,146 @@ func (w *SourceWorker) HandleError(ctx context.Context, req *pb.HandleWorkerErro
 
 	return st.HandleError(ctx, req, w.getRelayWithoutLock())
 }
+
+func (w *SourceWorker) observeValidatorStage(ctx context.Context, lastUsedRev int64) error {
+	var wg sync.WaitGroup
+
+	startRevision := lastUsedRev + 1
+	for {
+		stageCh := make(chan ha.Stage, 10)
+		errCh := make(chan error, 10)
+		wg.Add(1)
+		watchCtx, watchCancel := context.WithCancel(ctx)
+		go func() {
+			defer func() {
+				close(stageCh)
+				close(errCh)
+				wg.Done()
+			}()
+			ha.WatchValidatorStage(watchCtx, w.etcdClient, w.cfg.SourceID, startRevision, stageCh, errCh)
+		}()
+		err := w.handleValidatorStage(watchCtx, stageCh, errCh)
+		watchCancel()
+		wg.Wait()
+
+		if etcdutil.IsRetryableError(err) {
+			startRevision = 0
+			retryNum := 1
+			for startRevision == 0 {
+				select {
+				case <-ctx.Done():
+					return nil
+				case <-time.After(500 * time.Millisecond):
+					startRevision, err = w.getCurrentValidatorRevision(w.cfg.SourceID)
```

Comment:
we should also reset the validator stage to ones in etcd 🤔 example:

there are three historical stages: (running, rev1), (stopped, rev2), (running, rev3)
At first, the validator starts with subtask, it gets (running, rev1) and watch from rev1. Then it gets (stopped, rev2) and stops validator. At this time etcd compacts and cause CompactError, we enter `if etcdutil.IsRetryableError(err)` and now `getCurrentValidatorRevision` returns rev3, we start watch from rev3 but no KV changes now.

Now the stage in etcd is running but we haven't started validator

Reply:
didn't add 1 to revision in the loop so we can get the last one, see:

https://github.com/D3Hunter/tiflow/blob/47f24bbbc7252b9ff772ddab5129393fca0102a6/dm/dm/worker/source_worker.go#L1140